internal/transport/http2_server.go

```diff
@@ -1114,27 +1117,38 @@ func (t *http2Server) WriteStatus(s *Stream, st *status.Status) error {
 
 // Write converts the data into HTTP2 data frame and sends it out. Non-nil error
 // is returns if it fails (e.g., framing error, transport error).
-func (t *http2Server) Write(s *Stream, hdr []byte, data []byte, opts *Options) error {
+func (t *http2Server) Write(s *Stream, hdr []byte, data mem.BufferSlice, opts *Options) error {
+	reader := data.Reader()
+	data.Free()
```

Comment:
Talked about this offline. It's less error prone to have `.Reader` take a reference and guarantee that it will be there rather than having to make sure `Reader` has its own reference. The overall attitude towards buffers is that callers to a function manage the lifecycle of buffers, and if a function wants to guarantee that a buffer lives beyond the function returning, it should always take its own reference. This has been updated in the codec as well